src/pybind/mgr/cephadm/serve.py

```diff
@@ -667,3 +691,19 @@ def convert_tags_to_repo_digest(self) -> None:
                 image_info = digests[container_image_ref]
                 if image_info.repo_digest:
                     self.mgr.set_container_image(entity, image_info.repo_digest)
+
+    # ha-rgw needs definitve host list to create keepalived config files
+    # if definitive host list has changed, all ha-rgw daemons must get new
+    # config, including those that are already on the correct host and not
+    # going to be deployed
+    def update_ha_rgw_definitive_hosts(self, spec: ServiceSpec, hosts: List[HostPlacementSpec],
+                                       add_hosts: Set[HostPlacementSpec]) -> HA_RGWSpec:
+        spec = cast(HA_RGWSpec, spec)
+        if not (set(hosts) == set(spec.definitive_host_list)):
+            spec.definitive_host_list = hosts
+            ha_rgw_daemons = self.mgr.cache.get_daemons_by_service(spec.service_name())
+            for daemon in ha_rgw_daemons:
+                if daemon.hostname in [h.hostname for h in hosts] and daemon.hostname not in add_hosts:
+                    self.mgr.cache.schedule_daemon_action(
+                        daemon.hostname, daemon.name(), 'reconfig')
+        return spec
```

Comment:
I though this is why we added 

https://github.com/ceph/ceph/blob/d8675ecf803dca65be8e1af967cb81ba0061d309/src/pybind/mgr/cephadm/module.py#L426

?

Reply:
the definitive hosts are used in creating the keepalived config file (it needs to know where other keepalived are being placed) that it uses on deployment. That means, if done using post actions, every single keepalived daemon would need to be reconfigured right after it is initially deployed to get the config file with the location of the other daemons. Doing it the way it is here allows new keepalived daemons to get this info before they are created so you only have to reconfig already existing daemons if the set of hosts ha-rgw is on changes. It should still be functional using post actions but all new keepalived daemons would effectively have to be deployed twice. This is sort of an odd case as we would ideally like to set this attribute in the small window between when placement is done and when the daemons are actually created.